### PR TITLE
Add provider tests for ERC20

### PR DESCRIPTION
**Pull request by alexanderattar**
## Description

Adds tests for interacting with ERC20 tokens using the JsonRpcProvider and also stubs out test for withdraws on SimpleStorage

## Questions
- Can we interact with the contract using the JsonRpcProvider and Optimism Provider at the same time?

### Fixes
- Fixes https://github.com/ethereum-optimism/roadmap/issues/227

## Contributing Agreement

- [x] I have read and understood the [Optimism Contributing Guide and Code of Conduct](https://github.com/ethereum-optimism/optimism-monorepo/blob/master/.github/CONTRIBUTING.md) and am following those guidelines in this pull request.

### CODE REVIEW

File: packages/integration-tests/contracts/ERC20.sol
Comment (tynes) [inline packages/integration-tests/contracts/ERC20.sol]: What do you think about deduplicating the need to have two copies of the same contract by putting the contracts in the base of the repo as well as the waffle config?

File: packages/integration-tests/test/actions.spec.ts
Comment (tynes) [inline packages/integration-tests/test/actions.spec.ts]: This could be a `require` instead of `fs.readFileSync`, `require` on a json file will turn it into an object automatically

File: packages/integration-tests/test/actions.spec.ts
Comment (tynes) [inline packages/integration-tests/test/actions.spec.ts]: This could probably have a `await simpleStorage.wait()`

File: packages/integration-tests/test/actions.spec.ts
Comment (tynes) [inline packages/integration-tests/test/actions.spec.ts]: What do you think about breaking up this function `deployERC20` by having `erc20Json` and `ERC20Factory` be constants at the top of hte file and then just calling `ERC20Factory.deploy` in the `it` block

File: packages/integration-tests/test/actions.spec.ts
Comment (tynes) [inline packages/integration-tests/test/actions.spec.ts:107]: Great tests, easy to read

File: packages/integration-tests/test/actions.spec.ts
Comment (tynes) [inline packages/integration-tests/test/actions.spec.ts:147]: `receipt` is unused here, could assert on the `Transfer` log being present

File: packages/integration-tests/test/actions.spec.ts
Comment (tynes) [inline packages/integration-tests/test/actions.spec.ts]: Don't worry about this for this PR, just using the ethers `JsonRpcProvider` is fine

File: packages/integration-tests/test/actions.spec.ts
Comment (alexanderattar) [reply packages/integration-tests/test/actions.spec.ts]: Nice. Definitely cleaner

File: packages/integration-tests/test/actions.spec.ts
Comment (alexanderattar) [reply packages/integration-tests/test/actions.spec.ts]: 👍 good call

File: packages/integration-tests/test/actions.spec.ts
Comment (alexanderattar) [reply packages/integration-tests/test/actions.spec.ts]: 👍 sounds good

File: packages/integration-tests/test/actions.spec.ts
Comment (alexanderattar) [reply packages/integration-tests/test/actions.spec.ts]: done

File: packages/integration-tests/contracts/ERC20.sol
Comment (alexanderattar) [reply packages/integration-tests/contracts/ERC20.sol]: Good call

File: packages/integration-tests/test/actions.spec.ts
Comment (tynes) [inline packages/integration-tests/test/actions.spec.ts]: Is this meant to be disabled still?

File: packages/integration-tests/test/actions.spec.ts
Comment (tynes) [inline packages/integration-tests/test/actions.spec.ts]: Does this emit more than 1 event?

File: packages/integration-tests/test/actions.spec.ts
Comment (K-Ho) [inline packages/integration-tests/test/actions.spec.ts]: we sure this works? Just checked and it looks like etherbase is an address, not a private key?

File: packages/integration-tests/test/actions.spec.ts
Comment (K-Ho) [inline packages/integration-tests/test/actions.spec.ts]: Wouldn't the `Transfer` event emit a value of 100 not 0?

File: packages/integration-tests/test/actions.spec.ts
Comment (K-Ho) [inline packages/integration-tests/test/actions.spec.ts]:
```suggestion
    const senderBalance = await erc20.balanceOf(l2Wallet.address)
    expect(senderBalance.toNumber()).to.equal(INITIAL_AMOUNT - 100)
    const newBalance = await erc20.balanceOf(alice.address)
```

File: packages/tx-ingestion/package.json
Comment (K-Ho) [inline packages/tx-ingestion/package.json:11]: nice!

**Review by K-Ho — COMMENTED** [review]
Looks great! couple small nit fixes - also, not sure how you're able to create Alice's wallet with just an address :O Ping me once these small things are in and I'll approve!

File: packages/integration-tests/test/actions.spec.ts
Comment (alexanderattar) [reply packages/integration-tests/test/actions.spec.ts]: Yeah, when I talked to Kevin before the break it was noted to prioritize the transfer test and do withdrawal in another PR

File: packages/integration-tests/test/actions.spec.ts
Comment (alexanderattar) [reply packages/integration-tests/test/actions.spec.ts]: I'll make this change

File: packages/integration-tests/test/actions.spec.ts
Comment (alexanderattar) [reply packages/integration-tests/test/actions.spec.ts]: It returns two `Transfer` events. One with `logIndex: 0` and one with `logIndex: 1`. Do you have any ideas?

```
[ { transactionIndex: 0,
    blockNumber: 11,
    transactionHash:
     '0x1a6b0f3e8ecf334eeb1a7282db1c33975cf4b364069ecf3832621f0682ce48a0',
    address: '0x4200000000000000000000000000000000000006',
    topics:
     [ '0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef',
       '0x000000000000000000000000023ffdc1530468eb8c8eebc3e38380b5bc19cc5d',
       '0x0000000000000000000000004200000000000000000000000000000000000005' ],
    data:
     '0x0000000000000000000000000000000000000000000000000000000000000000',
    logIndex: 0,
    blockHash:
     '0x45a6eda785c8e056c6f1fbd123d3d03f2820f2fcd35ae4938bea421de87656e0',
    args:
     [ '0x023fFdC1530468eb8c8EEbC3e38380b5bc19Cc5d',
       '0x4200000000000000000000000000000000000005',
       [BigNumber],
       _from: '0x023fFdC1530468eb8c8EEbC3e38380b5bc19Cc5d',
       _to: '0x4200000000000000000000000000000000000005',
       _value: [BigNumber] ],
    decode: [Function],
    event: 'Transfer',
    eventSignature: 'Transfer(address,address,uint256)',
    removeListener: [Function],
    getBlock: [Function],
    getTransaction: [Function],
    getTransactionReceipt: [Function] },
  { transactionIndex: 0,
    blockNumber: 11,
    transactionHash:
     '0x1a6b0f3e8ecf334eeb1a7282db1c33975cf4b364069ecf3832621f0682ce48a0',
    address: '0x6418E5Da52A3d7543d393ADD3Fa98B0795d27736',
    topics:
     [ '0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef',
       '0x000000000000000000000000023ffdc1530468eb8c8eebc3e38380b5bc19cc5d',
       '0x0000000000000000000000000a4984541f192294ca4cca6788519204b45d67ea' ],
    data:
     '0x0000000000000000000000000000000000000000000000000000000000000064',
    logIndex: 1,
    blockHash:
     '0x45a6eda785c8e056c6f1fbd123d3d03f2820f2fcd35ae4938bea421de87656e0',
    args:
     [ '0x023fFdC1530468eb8c8EEbC3e38380b5bc19Cc5d',
       '0x0A4984541f192294cA4cCa6788519204B45d67eA',
       [BigNumber],
       _from: '0x023fFdC1530468eb8c8EEbC3e38380b5bc19Cc5d',
       _to: '0x0A4984541f192294cA4cCa6788519204B45d67eA',
       _value: [BigNumber] ],
    decode: [Function],
    event: 'Transfer',
    eventSignature: 'Transfer(address,address,uint256)',
    removeListener: [Function],
    getBlock: [Function],
    getTransaction: [Function],
    getTransactionReceipt: [Function] } ]
```

Comment (K-Ho) [reply packages/integration-tests/test/actions.spec.ts]: One of the transfer events is the transfer of ETH to pay fee (this will be a transfer of value 0) see https://github.com/ethereum-optimism/contracts-v2/blob/d5a17efac0719a3645e41599272a28a099b1ee0b/contracts/optimistic-ethereum/OVM/accounts/OVM_ECDSAContractAccount.sol#L90-L95

File: packages/integration-tests/test/actions.spec.ts
Comment (alexanderattar) [reply packages/integration-tests/test/actions.spec.ts]: Yeah you're right. It was related to https://github.com/ethereum-optimism/integration-tests/pull/36#discussion_r551605983. First event pays fee of 0 and the second event has the value of 100. Modifying the test.

File: packages/integration-tests/test/actions.spec.ts
Comment (alexanderattar) [reply packages/integration-tests/test/actions.spec.ts]: It does work the theory being that this is one of the default addresses so maybe the pk is queryable? We could try using a different new test pk too. Do you guys have one in mind that is not the deployer pk?

File: packages/integration-tests/test/actions.spec.ts
Comment (alexanderattar) [reply packages/integration-tests/test/actions.spec.ts]: The problem was that `.find` only returns the first element that matches https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find. 

`.filter` might be a more appropriate function to use, but both events are `Transfer` so we don't really gain anything from using that.

File: packages/integration-tests/test/actions.spec.ts
Comment (tynes) [inline packages/integration-tests/test/actions.spec.ts]: What is this sleep for?

File: packages/integration-tests/test/actions.spec.ts
Comment (tynes) [inline packages/integration-tests/test/actions.spec.ts:149]: Good comment. Eventually we will want to remove the event for the fee because users will not expect it, so in the future that change will break this test

Comment (tynes) [reply packages/integration-tests/test/actions.spec.ts:149]: We should assert there are only 2 events

Comment (alexanderattar) [reply packages/integration-tests/test/actions.spec.ts:149]: Good call. I'll add a check for that

File: packages/integration-tests/test/actions.spec.ts
Comment (alexanderattar) [reply packages/integration-tests/test/actions.spec.ts]: I chatted with @K-Ho and we're going to remove that and implement the withdraw function using the watcher instead. We discussed leaving it as is at the moment since this is disabled and then I'm going to open a new PR for finishing the withdrawal test in the next iteration.

Comment (tynes) [reply packages/integration-tests/test/actions.spec.ts]: We should be careful about getting in the habit of adding dead code to the codebase because of the risk of something else coming up and the dead code never being removed. In this case its the `watcher` and `initWatcher` as well as the skipped withdrawal tests

Comment (alexanderattar) [reply packages/integration-tests/test/actions.spec.ts]: The watcher and initWatcher is actually used by the deposit test, but I'll remove the withdraw function and withdrawal test for now and just re-add them to the branch I'm going to open for withdrawals

File: packages/integration-tests/test/actions.spec.ts
Comment (tynes) [reply packages/integration-tests/test/actions.spec.ts]: It should be safe to use the deployer private key here since there won't be a race condition. I'm a bit confused as well how this works if you pass in an address, it looks like this `Wallet` is used to deploy the erc20 contract

File: packages/tx-ingestion/test/sequencer-batch-append.spec.ts
Comment (tynes) [inline packages/tx-ingestion/test/sequencer-batch-append.spec.ts:13]: This file definitely needs to be built out so that it tests sequencer batch append but that is out of scope of this PR

**Review by tynes — COMMENTED** [review]
This PR looks good to me besides the wallet being created with the address confusion

File: packages/tx-ingestion/test/sequencer-batch-append.spec.ts
Comment (alexanderattar) [reply packages/tx-ingestion/test/sequencer-batch-append.spec.ts:13]: Got it. I'll keep that in mind

**Comment by tynes** [comment]
This PR looks good to me, will approve after CI passes

**Review by tynes — APPROVED** [review]

**Pull request merged**
